### PR TITLE
[New Feature]add lovasz loss

### PR DESCRIPTION
**Pull request by yamengxi**

### CODE REVIEW

**Comment by hellock** [comment]
Task linked: [CU-j6umcn Lovász-Softmax loss #351](https://app.clickup.com/t/j6umcn)

File: mmseg/models/losses/lovasz_loss.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
+import mmcv
```
Comment (xvjiarui) [inline mmseg/models/losses/lovasz_loss.py:1]: We may add sth like
```
Modified from ...
```

File: mmseg/models/losses/lovasz_loss.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+                 classes='present',
+                 per_image=False,
+                 class_weight=None,
+                 reduction='mean',
+                 avg_factor=None,
+                 ignore_index=255):
+    """Binary Lovasz hinge loss.
+
+    Args:
+        probas (torch.Tensor): [B, H, W], logits at each pixel
```
Comment (xvjiarui) [inline mmseg/models/losses/lovasz_loss.py:92]: `probas` is not an ideal abbr. 
We may use `logits` directly.

Comment (xvjiarui) [reply mmseg/models/losses/lovasz_loss.py:92]: In our codebase, `logits` usually refers to unnormalized energy. `prob` refers to probability distribution.

File: mmseg/models/losses/lovasz_loss.py [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+        loss = lovasz_hinge_flat(
+            *flatten_binary_scores(probas, labels, ignore_index))
+    return loss
+
+
+def lovasz_softmax_flat(probas, labels, classes='present', class_weight=None):
+    """Multi-class Lovasz-Softmax loss.
+
+    Args:
+        probas (torch.Tensor): [P, C], class probabilities at each prediction
```
Comment (xvjiarui) [inline mmseg/models/losses/lovasz_loss.py:130]: Similarly.

File: mmseg/models/losses/lovasz_loss.py [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+                   classes='present',
+                   per_image=False,
+                   class_weight=None,
+                   reduction='mean',
+                   avg_factor=None,
+                   ignore_index=255):
+    """Multi-class Lovasz-Softmax loss.
+
+    Args:
+        probas (torch.Tensor): [B, C, H, W], class probabilities at each
```
Comment (xvjiarui) [inline mmseg/models/losses/lovasz_loss.py:180]: Similarly.

File: mmseg/models/losses/lovasz_loss.py [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+    return loss
+
+
+@LOSSES.register_module()
+class LovaszLoss(nn.Module):
+    """LovaszLoss.
+
+    This loss is proposed in `The Lovasz-Softmax loss: A tractable surrogate
+    for the optimization of the intersection-over-union measure in neural
+    networks <https://openaccess.thecvf.com/content_cvpr_2018/html/Berman_The_LovaSz-Softmax_Loss_CVPR_2018_paper.html>`_. # noqa
```
Comment (xvjiarui) [inline mmseg/models/losses/lovasz_loss.py:228]: `https://arxiv.org/abs/1705.08790` should be fine.

File: mmseg/models/losses/lovasz_loss.py [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+
+
+def lovasz_softmax_flat(probas, labels, classes='present', class_weight=None):
+    """Multi-class Lovasz-Softmax loss.
+
+    Args:
+        probas (torch.Tensor): [P, C], class probabilities at each prediction
+            (between 0 and 1).
+        labels (torch.Tensor): [P], ground truth labels (between 0 and C - 1).
+        classes (str | list[int], optional): Classes choosed to calculate loss.
```
Comment (xvjiarui) [inline mmseg/models/losses/lovasz_loss.py:133]: Add
```options are 'all', 'present'. ```

Comment (xvjiarui) [reply mmseg/models/losses/lovasz_loss.py:133]: Check if the argument is valid.

File: mmseg/models/losses/lovasz_loss.py [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+        labels (torch.Tensor): [P], binary ground truth labels (0 or 1).
+
+    Returns:
+        torch.Tensor: The calculated loss.
+    """
+    if len(labels) == 0:
+        # only void pixels, the gradients should be 0
+        return logits.sum() * 0.
+    signs = 2. * labels.float() - 1.
+    errors = (1. - logits * Variable(signs))
```
Comment (xvjiarui) [inline mmseg/models/losses/lovasz_loss.py:72]: Why `Variable` here?

File: mmseg/models/losses/lovasz_loss.py (hunk 1 above)
Comment (yamengxi) [reply mmseg/models/losses/lovasz_loss.py:1]: Get it!

File: mmseg/models/losses/lovasz_loss.py (hunk 7 above)
Comment (yamengxi) [reply mmseg/models/losses/lovasz_loss.py:75]: I don't know, it is the original code.

Comment (yamengxi) [reply mmseg/models/losses/lovasz_loss.py:75]: https://github.com/bermanmaxim/LovaszSoftmax/blob/master/pytorch/lovasz_losses.py#L105

**Review by xvjiarui — APPROVED** [review]
